File: cpp/stack/nearestSmallerElement.cpp
#include <bits/stdc++.h>
using namespace std;

void printArray(int arr[], int n){
    for(int i=0; i<n; i++)
        cout << arr[i] << " ";
}

void printVector(vector<int> v) {
    for(auto elem : v){
        cout << elem << " ";
    }
}

void nearestSmallerElementRight (int arr[], int n) {
    vector<int> output;
    stack<int> s;

    for(int i=n-1; i>=0; i--) {
        while(s.size() > 0 && s.top() >= arr[i])
            s.pop();
        if (s.size() == 0)
            output.push_back(-1);
        else
            output.push_back(s.top());

        s.push(arr[i]);
    }
    printVector(output);
}

void nearestSmallerElementLeft (int arr[], int n) {
    vector<int> output;
    stack<int> s;

    for(int i=0; i<n; i++) {
        while(s.size() > 0 && s.top() >= arr[i])
            s.pop();
        if (s.size() == 0)
            output.push_back(-1);
        else
            output.push_back(s.top());

        s.push(arr[i]);
    }
    printVector(output);
}

int main(){
    int arr[] = {10, 5, 11, 6, 20, 12};
    int n = sizeof(arr) / sizeof(arr[0]);

    cout << "The next smaller elements of ";
    printArray(arr, n);
    cout << "to right are ";
    nearestSmallerElementRight(arr, n);

    cout << "\nThe next smaller elements of ";
    printArray(arr, n);
    cout << "to left are ";
    nearestSmallerElementLeft(arr, n);
}

File: Stack/nearestSmallerElement.py
def nearestSmallerElementRight(arr):
    stack = []
    output = [None] * n

    for i in range(n-1, -1, -1):
        while (len(stack)!=0) and (arr[i] <= stack[-1]):
            stack.pop()
        if len(stack)==0:
            output[i] = -1
        else:
            output[i] = stack[-1]
        stack.append(arr[i])
    
    return output

def nearestSmallerElementLeft(arr):
    stack = []
    output = [None] * n

    for i in range(n):
        while (len(stack)!=0) and (arr[i] <= stack[-1]):
            stack.pop()
        if len(stack)==0:
            output[i] = -1
        else:
            output[i] = stack[-1]
        stack.append(arr[i])
    
    return output

arr = [10, 5, 11, 6, 20, 12]
n = len(arr)

ngeRight = nearestSmallerElementRight(arr)
ngeLeft = nearestSmallerElementLeft(arr)

print("The next greatest elements of {} to right are {}".format(arr, ngeRight))
print("The next greatest elements of {} to left are {}".format(arr, ngeLeft))